src/block_cache.rs

```diff
@@ -0,0 +1,390 @@
+    /// length of `read_buf` is `max_blocks_per_read * block_size`.
+    max_blocks_per_read: u32,
+
+    /// Cache entries, sorted from most-recently-used to least.
+    ///
+    /// The entries are fully allocated when the cache is
+    /// created. During regular operation no additional allocation or
+    /// deallocation occurs, data is just copied around.
+    entries: VecDeque<CacheEntry>,
+
+    /// File system block size.
+    block_size: BlockSize,
+
+    /// Total number of blocks in the filesystem.
+    ///
+    /// This is used to ensure that when reading multiple blocks we
+    /// don't go past the end of the filesystem.
+    num_fs_blocks: u64,
+}
+
+impl BlockCache {
+    /// Create a block cache with sensible defaults.
+    pub(crate) fn new(
+        block_size: BlockSize,
+        num_fs_blocks: u64,
+    ) -> Result<Self, Ext4Error> {
+        Self::with_opts(CacheOpts::new(block_size), num_fs_blocks)
+    }
+
+    /// Create a block cache with control over the number of entries and
+    /// the read size.
+    ///
+    /// # Preconditions
+    ///
+    /// `max_blocks_per_read` must be less than or equal to `num_entries`.
+    fn with_opts(
+        opts: CacheOpts,
+        num_fs_blocks: u64,
+    ) -> Result<Self, Ext4Error> {
+        assert!(usize_from_u32(opts.max_blocks_per_read) <= opts.num_entries);
+
+        let read_buf_len = opts.read_buf_size_in_bytes();
+
+        let entries = vec![
+            CacheEntry {
+                block_index: 0,
+                data: vec![0; opts.block_size.to_usize()].into_boxed_slice(),
+            };
+            opts.num_entries
+        ];
+        Ok(Self {
+            entries: VecDeque::from(entries),
+            max_blocks_per_read: opts.max_blocks_per_read,
+            read_buf: vec![0; read_buf_len].into_boxed_slice(),
+            block_size: opts.block_size,
+            num_fs_blocks,
+        })
+    }
+
+    /// Get the number of blocks to read.
+    ///
+    /// Normally this returns `max_blocks_per_read`. If reading that
+    /// many blocks would go past the end of the filesystem, the number
+    /// is clamped to avoid that.
+    ///
+    /// # Preconditions
+    ///
+    /// `block_index` must be less than `num_fs_blocks`.
+    fn num_blocks_to_read(&self, block_index: FsBlockIndex) -> u32 {
+        assert!(block_index < self.num_fs_blocks);
+
+        // Get the index of the block right after the last block to read.
+        let end_block = block_index
+            .saturating_add(u64::from(self.max_blocks_per_read))
+            .min(self.num_fs_blocks);
+
+        // OK to unwrap: `end_block` can't be less than `block_index`.
+        let num_blocks = end_block.checked_sub(block_index).unwrap();
+
+        // OK to unwrap: the number is at most `max_blocks_per_read`,
+        // which is a `u32`.
+        u32::try_from(num_blocks).unwrap()
+    }
+
+    /// Get the cache entry for `block_size`, reading and inserting
```

Comment:
Get the cache entry for `block_size` -> `block_index`?

Reply:
done

Reply:
Or... not? I pushed the commit, but it's not showing up in the github UI.

Reply:
https://www.githubstatus.com/incidents/vxfd5br11t6v

Reply:
amended and forced pushed

Reply:
Says the incident is resolved, but I still see the 'Processing updates' spinner. Maybe re-push?